File: packages/shared-rules-engine/src/url.ts
/** Strip scheme and query; keep host + path. */
export function normalizeUrl(raw: string): string {
  let u = raw.trim().toLowerCase();
  u = u.replace(/^https?:\/\//, "");
  const q = u.indexOf("?");
  if (q >= 0) u = u.slice(0, q);
  const hash = u.indexOf("#");
  if (hash >= 0) u = u.slice(0, hash);
  return u.replace(/\/+$/, "") || u;
}

/** Longest matching pattern wins. */
export function matchUrlPattern(
  normalizedUrl: string,
  pattern: string,
): boolean {
  const p = pattern.toLowerCase().replace(/^https?:\/\//, "").replace(/\/+$/, "");

  if (p.startsWith("*.")) {
    const suffix = p.slice(1);
    const host = normalizedUrl.split("/")[0] ?? "";
    return host === suffix.slice(1) || host.endsWith(suffix);
  }

  if (p.includes("/")) {
    return normalizedUrl === p || normalizedUrl.startsWith(p);
  }

  const host = normalizedUrl.split("/")[0] ?? "";
  return host === p || host.endsWith("." + p);
}

export function patternSpecificity(pattern: string): number {
  return pattern.length;
}

File: apps/windows-agent/src/screen_time_rules/url.py
from __future__ import annotations


def normalize_url(raw: str) -> str:
    u = raw.strip().lower()
    if u.startswith("https://"):
        u = u[8:]
    elif u.startswith("http://"):
        u = u[7:]
    q = u.find("?")
    if q >= 0:
        u = u[:q]
    h = u.find("#")
    if h >= 0:
        u = u[:h]
    return u.rstrip("/") or u


def match_url_pattern(normalized_url: str, pattern: str) -> bool:
    p = pattern.lower().replace("https://", "").replace("http://", "").rstrip("/")

    if p.startswith("*."):
        suffix = p[1:]
        host = normalized_url.split("/")[0]
        return host == suffix[1:] or host.endswith(suffix)

    if "/" in p:
        return normalized_url == p or normalized_url.startswith(p)

    host = normalized_url.split("/")[0]
    return host == p or host.endswith("." + p)


def pattern_specificity(pattern: str) -> int:
    return len(pattern)
